L1Trigger/Phase2L1ParticleFlow/src/newfirmware/jetmet/L1SeedConePFJetEmulator.h

```diff
@@ -0,0 +1,231 @@
+    return (unsigned)y(NB - 1, 0);
+  }
+
+  template <class data_T, class table_T, int N>
+  void init_invert_table(table_T table_out[N]) {
+    // The template data_T is the data type used to address the table
+    for (unsigned i = 0; i < N; i++) {
+      float x = real_val_from_idx<data_T, N>(i);
+      table_T inv_x = 1 / x;
+      table_out[i] = inv_x;
+    }
+  }
+
+  template <class in_t, class table_t, int N>
+  table_t invert_with_shift(in_t in, bool debug = false) {
+    table_t inv_table[N];
+    init_invert_table<in_t, table_t, N>(inv_table);
+
+    // find the first '1' in the denominator
+    int msb = 0;
+    for (int b = 0; b < in.width; b++) {
+      if (in[b])
+        msb = b;
+    }
+    // shift up the denominator such that the left-most bit (msb) is '1'
+    in_t in_shifted = in << (in.width - msb - 1);
+    // lookup the inverse of the shifted input
+    int idx = idx_from_real_val<in_t, N>(in_shifted);
+    table_t inv_in = inv_table[idx];
+    // shift the output back
+    table_t out = inv_in << (in.width - msb - 1);
+    if (debug) {
+      std::cout << "           x " << in << ", msb = " << msb << ", shift = " << (in.width - msb) << ", idx = " << idx
+                << std::endl;
+      std::cout << "     pre 1 / " << in_shifted << " = " << inv_in << "(" << 1 / (float)in_shifted << ")" << std::endl;
+      std::cout << "    post 1 / " << in << " = " << out << "(" << 1 / (float)in << ")" << std::endl;
+    }
+    return out;
+  }
+
+  detaphi_t deltaPhi(Particle a, Particle b) {
+    detaphi_t dphi = detaphi_t(a.hwPhi) - detaphi_t(b.hwPhi);
+    // phi wrap
+    detaphi_t dphi0 = dphi > PI ? detaphi_t(TWOPI - dphi) : detaphi_t(dphi);
+    detaphi_t dphi1 = dphi < -PI ? detaphi_t(TWOPI + dphi) : detaphi_t(dphi);
+    detaphi_t dphiw = dphi > detaphi_t(0) ? dphi0 : dphi1;
+    return dphiw;
+  }
+
+  bool inCone(Particle seed, Particle part, detaphi2_t cone2, bool debug) {
+    // scale the particle eta, phi to hardware units
+    detaphi_t deta = detaphi_t(seed.hwEta) - detaphi_t(part.hwEta);
+    detaphi_t dphi = deltaPhi(seed, part);
+    bool ret = deta * deta + dphi * dphi < cone2;
+    //bool ret = r2 < cone2;
+    if (debug) {
+      detaphi2_t r2 = detaphi2_t(deta) * detaphi2_t(deta) + detaphi2_t(dphi) * detaphi2_t(dphi);
+      std::cout << "  part eta, seed eta: " << part.hwEta << ", " << seed.hwEta << std::endl;
+      std::cout << "  part phi, seed phi: " << part.hwPhi << ", " << seed.hwPhi << std::endl;
+      std::cout << "  pt, deta, dphi, r2, cone2, lt: " << part.hwPt << ", " << deta << ", " << dphi << ", " << deta * deta + dphi * dphi << ", "
+                << cone2 << ", " << ret << std::endl;
+    }
+    return ret;
+  }
+
+Jet makeJet_HW(const std::vector<Particle>& parts, Config config) {
+  // Seed Cone Jet algorithm with ap_fixed types and hardware emulation
+  using namespace L1SCJetEmu;
+  Particle seed = parts.at(0);
+
+  // Fine unless we start using saturation, in which case order matters
```

Comment:
our `pt_t` type does have saturation enabled, but if it's implemented correctly by the library I hope it still doesn't matter in which order we sum things: it's all _unsigned_ numbers and so if the total would above the ceiling it should saturate at some point in the partial sum and stay saturated

Reply:
Right, actually this comment is left over from the implementation before moving to the other newfirmware data types, and I didn't have saturation. But you're right, since it's unsigned the order doesn't matter after all.